### PR TITLE
Add block glyph

### PR DESCRIPTION
**Pull request by bryevdv**
This PR is a minimal proposal to add a `Block` glyph.  

ref: https://discourse.bokeh.org/t/squarify-w-bokeh-for-treemap/8671

In the above support question, it occurred that we support all the common combinations of specifying rectangular regions, *except one*:

* `rect` — x, y, width, height (center at x, y)
* `vbar` — top, bottom, width, x (center on x)
* `hbar` — left, right, height, y (center on y)
* `quad` — top, bottom, left, right (both corners)

The one we don't have is *x, y, width, height (**one corner** at x, y)*.  This is a common way to want to specify rectangular regions, as evidenced by the Discourse question demonstrating that another tool `squarify` generates such, and futher, that MPL just accepts that as input. 

This PR adds a `Block` glyph with those missing semantics. (Note: I would have preferred to call this glyph `Box` but we unfortunately used up that name on a layout already). Additionally:

* Moves internal `box.ts` and class to `lrtb.ts`
* Removes some unused attrs found in the bar glyphs

With this initial implementation, a treemap based on `squarify` becomes simple:

```
import pandas as pd
from squarify import normalize_sizes, squarify
from bokeh.plotting import figure, show

data = dict(one=[50], two= [25], three = [12], four = [6], five = [3])
df = pd.DataFrame.from_dict(data=data, orient='index', columns=['amount' ])
df.index.name = 'labels'

x, y = 0, 0
width, height = 600, 400

df.sort_values(by='amount', inplace=True, ascending=False)
df['rect_norm'] = normalize_sizes(df.amount, width, height)
rect_df = pd.DataFrame.from_dict(data=squarify(df.rect_norm, x, y, width, height))
rect_df.set_index(df.index, inplace=True)

df = df.join(rect_df, how='left')

p = figure(height=height, width=width, tooltips="@labels: @amount")
p.x_range.range_padding = p.y_range.range_padding = 0

p.block('x', 'y', 'dx', 'dy', source=df, fill_alpha=0.4)
show(p)
```

Screen Shot 2021-11-28 at 15 06 55

---

I'd like to hear thoughts on proceeding to finish work to add this glyph. Assuming there is agreement (or at least, enough support) I would move on to adding tests and documentation.

cc @bokeh/dev

### CODE REVIEW

**Comment by bryevdv** [comment]
As an aside, I had the old `pre-push` hook installed for years to prevent accidental pushes to main, and it worked great. But I kept forgetting to install the new hooks updated for our new branch structure, and have made a hash on `branch-3.0`. Apologies for any inconvenience (and `pre-push` is now installed locally)

**Comment by p-himik** [comment]
Wouldn't it be easier and also less confusing to users to simply add a new parameter to `rect`? Like e.g. `origin` which can be `'top-left'` or `'center'`, similar to `text_align` of `text`.

**Comment by bryevdv** [comment]
@p-himik I don't think so. Extra cash be rotated so they don't share the same base as quad and bar. They see are already more complicated. I also prefer separate API for the separate cases so they can be summarized in docs in a table.

**Comment by bryevdv** [comment]
>  take very similar args: translation ( x, y ) and scale ( width, height ).

I added a late edit (bad habit) but just to reiterate, I do think they are *not* the same. Rect scales cannot be negative (makes no sense at all, really) but allowing negative width/height values make sense for bars and would make sense here, too (and would also be useful to support). Have width/height mean different things (accept different values) in the same glyph, depending on the value of some other flag is impossible to document in a simple way. I think that is just a recipe for user confusion.

**Comment by g-parki** [comment]
(Down the road) the docs for this should probably have an aside about `match_aspect=True` since this requires data-space units to be truly accurate.

**Comment by bryevdv** [comment]
@g-parki I don't see how that is true? Aspect matching is important when the dimensions are not separable (e.g. a circle with one radius, have to pick on dimensions to "measure" along). But that's the the case here. Width is always measured on the x-axis and is correct along the x-axis, and similarly for y as well.

**Comment by bryevdv** [comment]
@g-parki or maybe I misunderstood the comment. But this glyph will *only* accept data units, if that is the concern.

**Comment by g-parki** [comment]
@bryevdv sorry I didn't articulate what I meant. I was speaking more toward the potential user expectation of having a true, non-dilated shape. You're right the actual data representation isn't at stake here.

**Comment by g-parki** [comment]
> Have width/height mean different things (accept different values) in the same glyph, depending on the value of some other flag is impossible to document in a simple way.

@bryevdv I hear you there. I think the least-intrusive way it could be done when the `rect` is center-anchored is by throwing a warning if any `width` or `height` are negative, and defaulting to the absolute value of each. My opinion is _this is acceptable_, but I totally respect your position on it. 😃

**Comment by bryevdv** [comment]
OK I haven't heard any strong objection or concern, there are a few 👍 and I definitely think it is a good idea personally, so I will be continuing work to finish this.

File: bokeh/models/glyphs.py [hunk 1]
```diff
@@ -349,6 +350,43 @@ class Bezier(LineGlyph):
     The {prop} values for the Bezier curves.
     """)
 
+class Block(LineGlyph, FillGlyph, HatchGlyph):
+    ''' Render rectangular regions, given a corner coordinate, width and height.
```
Comment (tcmetzger) [inline bokeh/models/glyphs.py:354]: I think we usually use Oxford commas, so I'd go
```suggestion
    ''' Render rectangular regions, given a corner coordinate, width, and height.
```

**Comment by bryevdv** [comment]
I wanted to add basic and nested treemap examples but will need to source some new hierarchical sampledata suitable for it. I am not sure I will have time in the immediate future but this is otherwise ready I would like to review and merge now and leave new examples for later. cc @bokeh/dev

**Comment by bryevdv** [comment]
Just noting, I will plan to merge this Friday, if there are not comments (or request for more time)

**Pull request merged**